rust/lance-index/src/reuse.rs

```diff
@@ -0,0 +1,421 @@
+        })
+    }
+}
+
+/// An index that stores row ID maps.
+/// A row ID map describes the mapping from old row address to new address after compactions.
+/// Each version contains the mapping for one round of compaction.
+#[derive(Debug, Clone)]
+pub struct FragReuseIndex {
+    pub details: FragReuseIndexDetails,
+
+    // row ID map of all compaction rounds
+    // ordered from oldest to newest
+    pub row_id_maps: Vec<HashMap<u64, Option<u64>>>,
+
+    store: Arc<dyn IndexStore>,
+}
+
+impl DeepSizeOf for FragReuseIndex {
+    fn deep_size_of_children(&self, context: &mut deepsize::Context) -> usize {
+        self.details.deep_size_of_children(context) + self.store.deep_size_of_children(context)
+    }
+}
+
+impl FragReuseIndex {
+    pub fn new(
+        row_id_maps: Vec<HashMap<u64, Option<u64>>>,
+        details: FragReuseIndexDetails,
+        store: Arc<dyn IndexStore>,
+    ) -> Self {
+        Self {
+            row_id_maps,
+            details,
+            store,
+        }
+    }
+
+    pub async fn load(
+        details: FragReuseIndexDetails,
+        store: Arc<dyn IndexStore>,
+    ) -> Result<Arc<Self>> {
+        let row_id_maps = join_all(
+            details
+                .versions
+                .iter()
+                .map(|version| load_row_id_map(&version.row_id_map_path, store.clone())),
+        )
+        .await
+        .into_iter()
+        .collect::<Result<Vec<_>>>()?;
+
+        Ok(Arc::new(Self::new(row_id_maps, details, store)))
+    }
+}
+
+#[derive(Serialize)]
+struct FragReuseStatistics {
+    num_versions: usize,
+    num_rows: usize,
+}
+
+#[async_trait]
+impl Index for FragReuseIndex {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_index(self: Arc<Self>) -> Arc<dyn Index> {
+        self
+    }
+
+    fn as_vector_index(self: Arc<Self>) -> Result<Arc<dyn crate::vector::VectorIndex>> {
+        Err(Error::NotSupported {
+            source: "FragReuseIndex is not a vector index".into(),
+            location: location!(),
+        })
+    }
+
+    fn statistics(&self) -> Result<serde_json::Value> {
+        let stats = FragReuseStatistics {
+            num_versions: self.details.versions.len(),
+            num_rows: self.details.versions.len(),
+        };
+        serde_json::to_value(stats).map_err(|e| Error::Internal {
+            message: format!("failed to serialize fragment reuse index statistics: {}", e),
+            location: location!(),
+        })
+    }
+
+    async fn prewarm(&self) -> Result<()> {
+        // TODO: can preload the row ID map
+        Ok(())
+    }
+
+    fn index_type(&self) -> IndexType {
+        IndexType::FragmentReuse
+    }
+
+    async fn calculate_included_frags(&self) -> Result<RoaringBitmap> {
+        unimplemented!()
+    }
+}
+
+/// Save a row ID map to a Lance index file.
+pub async fn save_row_id_map(
+    index_store: Arc<dyn IndexStore>,
+    dataset_version: &u64,
+    row_id_map: &HashMap<u64, Option<u64>>,
```

Comment:
This structure feels a bit unfortunate given that the common case is mapping sorted contiguous ranges to other sorted contiguous ranges. Have you considered re-using some of the `U64Segment` or `RowIdSequence` machinery?

Reply:
Yeah I used `HashMap` becuase these row ID maps come from the `RewriteResult` of `optimize.rewrite_files`, at that point it's already a hashmap and the sorting information is already lost. But I can make this storage more efficient with the data structure you mentioned, and then make the upstream changes in `optimize` based on what we arrive at here.